### ISSUE TITLE
Memory leaks in Tempesta

### ISSUE THREAD

**Issue by keshonok** [tempesta-tech/tempesta#691]
The setup:
* Tempesta's current master at 56b0ecb;
* Tempesta kernel 4.8.15  at [5b3aa7f](https://github.com/tempesta-tech/linux-4.8.15-tfw/commit/5b3aa7fe30ce46bb8f112a9544548da926cb3714);
* Both Tempesta and Nginx run on the same physical test server with 12 CPU cores and 64Gb RAM;
* Simplest Tempesta configuration:
```
server 127.0.0.1:8080
```

The configuration means that Tempesta bypasses cache on all requests. All requests are forwarded to the back-end server without exceptions.

The test is a simple `wrk -t 8 -c 1024 -d 600s http://192.168.0.1`. It runs on another physical machine with 8 CPU cores and 32Gb RAM. It is connected with the test server via 10G interface.

After each run of the above test the main server with Tempesta loses several gigabytes of RAM. Eventually all RAM on the main test server is exhausted and OOM is triggered, which effectively makes the test server unusable.

_Please note that the issue doesn't appear to present itself in a setup where Tempesta and Nginx run in the same VM, and the test is run on the host machine against Tempesta in the VM._

This is the final state of RAM on the test machine at the time of OOM. It also shows an approximate time it took to exhaust all memory.
```
top - 10:32:16 up 49 min,  1 user,  load average: 2.64, 2.11, 1.40
Tasks: 198 total,   1 running, 197 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us,  1.2 sy,  0.0 ni, 97.8 id,  0.4 wa,  0.0 hi,  0.6 si,  0.0 st
KiB Mem:  65326760 total, 64993280 used,   333480 free,      520 buffers
KiB Swap:  7998460 total,  2595384 used,  5403076 free.     1456 cached Mem
```

This shows the OOM triggered:
```
[  283.701873] [tdb] Opened table /opt/tempesta/db/filter.tdb: size=16777216 rec_size=20 base=ffff88103e000000
[  283.910192] [tdb] Opened table /opt/tempesta/db/cache.tdb: size=268435456 rec_size=0 base=ffff88102e000000
[  283.919917] [tempesta] Open listen socket on: 0.0.0.0
[  283.925812] [tempesta] modules are started
[ 2949.269612] [sync_sockets] Warning: Unable to copy an egress SKB.
[ 2949.270234] [sync_sockets] Warning: Unable to copy an egress SKB.
[ 2949.270417] [sync_sockets] Warning: Unable to copy an egress SKB.
[ 2949.270623] [sync_sockets] Warning: Unable to copy an egress SKB.
[ 2949.270630] [sync_sockets] Warning: Unable to copy an egress SKB.
[ 2949.270834] [sync_sockets] Warning: Unable to copy an egress SKB.
[ 2949.270969] [sync_sockets] Warning: Unable to copy an egress SKB.
[ 2949.271014] [sync_sockets] Warning: Unable to copy an egress SKB.
[ 2949.271098] [sync_sockets] Warning: Unable to copy an egress SKB.
[ 2949.271151] [sync_sockets] Warning: Unable to copy an egress SKB.
[ 2949.291179] TCP: request_sock_TCP: Possible SYN flooding on port 80. Dropping request.  Check SNMP counters.
[ 2950.262468] systemd invoked oom-killer: gfp_mask=0x24201ca(GFP_HIGHUSER_MOVABLE|__GFP_COLD), order=0, oom_score_adj=0
[ 2950.273099] systemd cpuset=/ mems_allowed=0
[ 2950.277342] CPU: 10 PID: 1 Comm: systemd Tainted: G           O    4.8.15-debug-ab+ #6
[ 2950.285268] Hardware name: Supermicro Super Server/X10SRH-CF, BIOS 2.0 12/17/2015
[ 2950.292757]  0000000000000286 00000000c03e5379 ffff8810182b3a20 ffffffff813167c5
[ 2950.300219]  ffff8810182b3bf8 ffff881013ad8000 ffff8810182b3a98 ffffffff81213905
[ 2950.307691]  ffffffff81c54a20 ffff8810182b3a48 ffffffff810c4bdd ffff8810182b3a68
[ 2950.315154] Call Trace:
[ 2950.317607]  [<ffffffff813167c5>] dump_stack+0x68/0x93
[ 2950.322754]  [<ffffffff81213905>] dump_header+0x5c/0x22e
[ 2950.328077]  [<ffffffff810c4bdd>] ? trace_hardirqs_on+0xd/0x10
[ 2950.333918]  [<ffffffff81185112>] oom_kill_process+0x212/0x3e0
[ 2950.339759]  [<ffffffff810782f7>] ? has_capability_noaudit+0x17/0x20
[ 2950.346122]  [<ffffffff81185854>] out_of_memory+0x524/0x540
[ 2950.351701]  [<ffffffff8118551e>] ? out_of_memory+0x1ee/0x540
[ 2950.357458]  [<ffffffff8118b5a8>] __alloc_pages_nodemask+0xf28/0xf60
[ 2950.363821]  [<ffffffff811e2ce1>] alloc_pages_current+0xa1/0x1d0
[ 2950.370203]  [<ffffffff8117f1e9>] __page_cache_alloc+0x129/0x150
[ 2950.376590]  [<ffffffff811803ec>] ? pagecache_get_page+0x2c/0x250
[ 2950.383070]  [<ffffffff81182c6d>] filemap_fault+0x32d/0x600
[ 2950.389030]  [<ffffffff81182da3>] ? filemap_fault+0x463/0x600
[ 2950.395168]  [<ffffffffa0156ba6>] ext4_filemap_fault+0x36/0x50 [ext4]
[ 2950.402004]  [<ffffffff811bb657>] __do_fault+0x87/0x160
[ 2950.407615]  [<ffffffff811bfc73>] handle_mm_fault+0x1003/0x1430
[ 2950.413925]  [<ffffffff811becc5>] ? handle_mm_fault+0x55/0x1430
[ 2950.420239]  [<ffffffff8105615a>] __do_page_fault+0x1fa/0x4e0
[ 2950.426371]  [<ffffffff8105644c>] do_page_fault+0xc/0x10
[ 2950.432068]  [<ffffffff815fbc32>] page_fault+0x22/0x30
[ 2950.437603] Mem-Info:
[ 2950.440738] active_anon:0 inactive_anon:23 isolated_anon:0
[ 2950.440738]  active_file:90 inactive_file:0 isolated_file:0
[ 2950.440738]  unevictable:0 dirty:1 writeback:32 unstable:0
[ 2950.440738]  slab_reclaimable:3954 slab_unreclaimable:17272
[ 2950.440738]  mapped:0 shmem:5 pagetables:2310 bounce:0
[ 2950.440738]  free:80251 free_pcp:184 free_cma:0
[ 2950.476886] Node 0 active_anon:0kB inactive_anon:76kB active_file:148kB inactive_file:56kB unevictable:0kB isolated(anon):0kB isolated(file):0kB mapped:0kB dirty:4kB writeback:128kB shmem:0kB shmem_thp: 0kB shmem_pmdmapped: 0kB anon_thp: 20kB writeback_tmp:0kB unstable:0kB pages_scanned:2259 all_unreclaimable? yes
[ 2950.508068] Node 0 DMA free:15880kB min:16kB low:28kB high:40kB active_anon:0kB inactive_anon:0kB active_file:0kB inactive_file:0kB unevictable:0kB writepending:0kB present:15972kB managed:15888kB mlocked:0kB slab_reclaimable:0kB slab_unreclaimable:8kB kernel_stack:0kB pagetables:0kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB
[ 2950.541536] lowmem_reserve[]: 0 1296 63755 63755
[ 2950.547643] Node 0 DMA32 free:250700kB min:1372kB low:2696kB high:4020kB active_anon:40kB inactive_anon:104kB active_file:156kB inactive_file:152kB unevictable:0kB writepending:0kB present:1965536kB managed:1352772kB mlocked:0kB slab_reclaimable:32kB slab_unreclaimable:552kB kernel_stack:48kB pagetables:288kB bounce:0kB free_pcp:240kB local_pcp:0kB free_cma:0kB
[ 2950.584171] lowmem_reserve[]: 0 0 62459 62459
[ 2950.590230] Node 0 Normal free:53856kB min:66192kB low:130148kB high:194104kB active_anon:0kB inactive_anon:0kB active_file:116kB inactive_file:100kB unevictable:0kB writepending:0kB present:65011712kB managed:63958100kB mlocked:0kB slab_reclaimable:15784kB slab_unreclaimable:69008kB kernel_stack:4176kB pagetables:8952kB bounce:0kB free_pcp:528kB local_pcp:120kB free_cma:0kB
[ 2950.628359] lowmem_reserve[]: 0 0 0 0
[ 2950.633771] Node 0 DMA: 0*4kB 1*8kB (U) 0*16kB 0*32kB 2*64kB (U) 1*128kB (U) 1*256kB (U) 0*512kB 1*1024kB (U) 1*2048kB (M) 3*4096kB (M) = 15880kB
[ 2950.650356] Node 0 DMA32: 87*4kB (UM) 74*8kB (UM) 60*16kB (UM) 33*32kB (UME) 27*64kB (UME) 14*128kB (ME) 4*256kB (ME) 3*512kB (UME) 8*1024kB (UME) 6*2048kB (UME) 54*4096kB (M) = 250700kB
[ 2950.670416] Node 0 Normal: 489*4kB (ME) 298*8kB (UME) 241*16kB (MEH) 558*32kB (UME) 149*64kB (UME) 50*128kB (UME) 22*256kB (UME) 8*512kB (UM) 2*1024kB (M) 0*2048kB 0*4096kB = 53764kB
[ 2950.689703] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=1048576kB
[ 2950.699738] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
[ 2950.699740] 166 total pagecache pages
[ 2950.699741] 34 pages in swap cache
[ 2950.699743] Swap cache stats: add 842961, delete 842927, find 174110/294022
[ 2950.699744] Free swap  = 5402536kB
[ 2950.699745] Total swap = 7998460kB
[ 2950.699961] 16748305 pages RAM
[ 2950.699962] 0 pages HighMem/MovableOnly
[ 2950.699963] 416615 pages reserved
[ 2950.699963] 0 pages hwpoisoned
[ 2950.699976] [ pid ]   uid  tgid total_vm      rss nr_ptes nr_pmds swapents oom_score_adj name
[ 2950.700022] [  279]     0   279     8233        0      20       3       77             0 systemd-journal
[ 2950.700028] [  281]     0   281     9992        1      21       3      137         -1000 systemd-udevd
[ 2950.700064] [  686]     0   686     6883        0      18       3       78             0 cron
[ 2950.700086] [  692]     0   692    13795        0      32       3      178         -1000 sshd
[ 2950.700089] [  704]     0   704     4953        0      16       3       77             0 systemd-logind
[ 2950.700093] [  771]   107   771    10527        0      25       3       97          -900 dbus-daemon
[ 2950.700108] [  777]   104   777     8346        0      21       3      169             0 ntpd
[ 2950.700113] [  781]     0   781    64663        0      27       4      271             0 rsyslogd
[ 2950.700130] [  795]     0   795     1054        0       8       3       36             0 acpid
[ 2950.700133] [  823]     0   823    22795        0      42       3      306             0 nginx
[ 2950.700137] [  825]    33   825   102298        0     199       4    79814             0 nginx
[ 2950.700152] [  828]    33   828   102298        0     199       4    79806             0 nginx
[ 2950.700173] [  829]    33   829   108056        0     210       4    85579             0 nginx
[ 2950.700176] [  832]    33   832   102298        0     199       4    79808             0 nginx
[ 2950.700195] [  834]    33   834   102298        0     199       4    79816             0 nginx
[ 2950.700216] [  835]    33   835   102298        0     199       4    79823             0 nginx
[ 2950.700219] [  837]    33   837   102298        0     199       4    79822             0 nginx
[ 2950.700222] [  839]    33   839   102298        0     199       4    79820             0 nginx
[ 2950.700237] [  881]     0   881     3630        0      11       3       38             0 agetty
[ 2950.700241] [  882]     0   882     9039        0      22       3      156             0 master
[ 2950.700244] [  883]   106   883     9555        0      24       3      137             0 pickup
[ 2950.700259] [  884]   106   884     9567        0      23       3      138             0 qmgr
[ 2950.700262] [  885]     0   885     3574        0      11       3       35             0 agetty
[ 2950.700266] [  891]     0   891    19424        0      44       3      316             0 apache2
[ 2950.700281] [  894]    33   894    91716        0      73       3      390             0 apache2
[ 2950.700287] [  922]    33   922    91716        0      73       3      391             0 apache2
[ 2950.700303] [  957]     0   957    20681        0      44       3      240             0 sshd
[ 2950.700325] [  959]     0   959     6060        1      18       3      707             0 bash
[ 2950.700329] [ 3500]   106  3500    10611        0      25       3      170             0 tlsmgr
[ 2950.700332] [ 3544]     0  3544     6489        0      17       3      173             0 top
[ 2950.700348] [ 3768]   106  3768    11212        0      28       3      221             0 smtpd
```

**Comment by keshonok** [comment]
Please note that this issue report is actually related to Tempesta in the version prior to merging of #660, just to assert that the memory leak is NOT connected with the massive code changes introduced there. The memory leak was first noticed while debugging #660, and then proven to exist in the previous version of Tempesta as well.

**Comment by keshonok** [comment]
This issue is related to #693.

The changes made in attempt to close in on the cause of this issue are described in [the comment to #693](https://github.com/tempesta-tech/tempesta/issues/693#issuecomment-285710388).

These versions of Tempesta and the kernel were tested in the original setup for this issue. The tests did not reveal any noticeable memory leaks. The whole system worked stable for hours.

The results lead to the same conclusion: the real culprit is in the patch to the kernel 4.8.15 for Tempesta.

**Comment by keshonok** [comment]
Further tests with the 4.1.27 compatibility branch and 4.1.27 kernel show that there's a memory leak, not big but noticeable. The compatibility branch includes #660.

**Comment by keshonok** [comment]
Considering the recent Tempesta kernel patch https://github.com/tempesta-tech/linux-4.8.15-tfw/commit/0bf3cf888a6bc61da0919c5c989bec938bda6e4b, I have run a number of tests to see if the issue is still present. The tests run on the testbed consisting of two physical servers. The servers are exactly the same as those mentioned in the first message of this issue. The servers are connected via 10G ethernet link.

The setup:
* Tempesta's current master at fbbba57;
* Tempesta kernel 4.8.15  at [0bf3cf8](https://github.com/tempesta-tech/linux-4.8.15-tfw/commit/0bf3cf888a6bc61da0919c5c989bec938bda6e4b);
* Both Tempesta and Nginx run on the same physical test server with 12 CPU cores and 64Gb RAM;
* Simplest Tempesta configuration:
```
server 192.168.0.1:8080;
cache 0;
```

The configuration means that Tempesta bypasses cache on all requests. All requests are forwarded to the back-end server without exceptions.

The test is a simple `wrk -t 8 -c 4096 -d 600s http://192.168.0.1`. It runs on another physical machine with 8 CPU cores and 32Gb RAM. It is connected with the test server via 10G interface. Other tests runs were for 1 hour and even for 12 hours.

The major conclusion is that the massive memory leak where the system lost several gigabytes of RAM after each 10min test is now gone. Which means that the kernel patch was in the right direction.

However, the other conclusion is that the memory leak is still present. The scale is a bit smaller, but it's still very critical. Now it takes about 11-12 hours to exhaust 60Gb of free RAM and reach OOM that kills the server.

Below is a description of several representative tests along with data collected during those tests.

* `wrk -t 8 -c 4096 -d 1h http://192.168.0.1` - the test run for one hour. Various memory-related kernel data was collected at periodic rate of 15 seconds. They are attached below.

[proc_buddyinfo_1h_test.txt](https://github.com/tempesta-tech/tempesta/files/1031795/proc_buddyinfo_1h_test.txt)
[proc_meminfo_1h_test.txt](https://github.com/tempesta-tech/tempesta/files/1031798/proc_meminfo_1h_test.txt)
[proc_slabinfo_1h_test.txt](https://github.com/tempesta-tech/tempesta/files/1031796/proc_slabinfo_1h_test.txt)
[proc_vmstat_1h_test.txt](https://github.com/tempesta-tech/tempesta/files/1031797/proc_vmstat_1h_test.txt)
[proc_zoneinfo_1h_test.txt](https://github.com/tempesta-tech/tempesta/files/1031799/proc_zoneinfo_1h_test.txt)

Data collected in the above files show, on the surface, that there's the loss of memory pages at more or less consistent rate. So I built a variant of Tempesta kernel with support of `page_owner` option for future tests.

* `wrk -t 8 -c 4096 -d 10m http://192.168.0.1` - the test run for 10 minutes. The goal was to see the data collected by the `page_owner` feature. The result is in the attached file. The first file shows how the data look when the test completed but Tempesta is not unloaded yet. The second file shows how the data look when Tempesta is unloaded.

[sorted_page_owner__tempesta_10m_test_completed.txt](https://github.com/tempesta-tech/tempesta/files/1031812/sorted_page_owner__tempesta_10m_test_completed.txt)
[sorted_page_owner__tempesta_10m_test_unloaded.txt](https://github.com/tempesta-tech/tempesta/files/1031813/sorted_page_owner__tempesta_10m_test_unloaded.txt)

* `wrk -t 8 -c 4096 -d 12h http://192.168.0.1` - the test run for twelve hours, approximately. The goal was to reach OOM, and the goal has been reached. The test was stopped at about 10.5 hours time from the time it started to collect data before the system crashed due to OOM.

[proc_buddyinfo_OOM_test.txt](https://github.com/tempesta-tech/tempesta/files/1031824/proc_buddyinfo_OOM_test.txt)
[proc_meminfo_OOM_test.txt](https://github.com/tempesta-tech/tempesta/files/1031828/proc_meminfo_OOM_test.txt)
[proc_slabinfo_OOM_test.txt](https://github.com/tempesta-tech/tempesta/files/1031825/proc_slabinfo_OOM_test.txt)
[proc_vmstat_OOM_test.txt](https://github.com/tempesta-tech/tempesta/files/1031826/proc_vmstat_OOM_test.txt)
[proc_zoneinfo_OOM_test.txt](https://github.com/tempesta-tech/tempesta/files/1031827/proc_zoneinfo_OOM_test.txt)

Here's Tempesta stats at that time: [perfstat_almost_OOM.txt](https://github.com/tempesta-tech/tempesta/files/1031830/perfstat_almost_OOM.txt).

Here's `top` data at the time when the system is very close to OOM:  [top_OOM_test.txt](https://github.com/tempesta-tech/tempesta/files/1031832/top_OOM_test.txt)
And here's the data from `page_owner` kernel feature taken in this OOM test: [sorted_page_owner__OOM_test.txt](https://github.com/tempesta-tech/tempesta/files/1031835/sorted_page_owner__OOM_test.txt)


Finally, I run a test without Tempesta against standard Nginx. The setup is almost the same with the exclusion of Tempesta:
* Tempesta kernel 4.8.15  at [0bf3cf8](https://github.com/tempesta-tech/linux-4.8.15-tfw/commit/0bf3cf888a6bc61da0919c5c989bec938bda6e4b);
* Nginx runs on the physical test server with 12 CPU cores and 64Gb RAM;
* The test runs on another physical machine with 8 CPU cores and 32Gb RAM;
* Servers are connected via 10G ethernet link.

The test command was a similar `wrk -t 8 -c 4096 -d 12h http://192.168.0.1`.  Running the test for four hours showed the loss of about 16Gb of free RAM, so the rate of the loss of free RAM is somewhat slower than when Tempesta is in action. Here's data from `top` after about 4 hours: 
[top_nginx_4h_test.txt](https://github.com/tempesta-tech/tempesta/files/1032039/top_nginx_4h_test.txt).
**[page_owner data for this test will be added later today]**


To summarize this, in my opinion, all of these data points in the direction that the real culprit is somewhere is the kernel mechanisms (Tempesta kernel patch), rather than in Tempesta itself. Possibly it's somewhere around the SKB page allocator that is intensively used in both cases: with Tempesta and without Tempesta.

**Comment by krizhanovsky** [comment]
Just faced the same problem on current master, just shortly after wrk start:

        $ ./wrk -t 16 -c 16384 -d 1h http://debian:80/

        [ 5053.824819] bash invoked oom-killer: gfp_mask=0x24200ca(GFP_HIGHUSER_MOVABLE), nodemask=0, order=0, oom_score_adj=0
        [ 5053.832895] bash cpuset=/ mems_allowed=0
        [ 5053.833957] CPU: 0 PID: 761 Comm: bash Tainted: G           O    4.9.35 #13
        [ 5053.835552] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS 1.9.1-1.fc24 04/01/2014
        [ 5053.836873]  ffffabcd8063b948 ffffffffb1b069f5 ffffabcd8063bb38 ffff9588d4d2e540
        [ 5053.836873]  ffffabcd8063b9d0 ffffffffb19c6f3a 0000000000000000 0000000000000000
        [ 5053.836873]  ffffffffb2243360 ffffabcd8063b980 ffffffffb18ad75d ffffabcd8063b9a0
        [ 5053.836873] Call Trace:
        [ 5053.836873]  [<ffffffffb1b069f5>] dump_stack+0x68/0x93
        [ 5053.836873]  [<ffffffffb19c6f3a>] dump_header+0x76/0x24a
        [ 5053.836873]  [<ffffffffb18ad75d>] ? trace_hardirqs_on+0xd/0x10
        [ 5053.836873]  [<ffffffffb194bcde>] oom_kill_process+0x1fe/0x3e0
        [ 5053.836873]  [<ffffffffb194c1a0>] out_of_memory+0x110/0x520
        [ 5053.836873]  [<ffffffffb194c261>] ? out_of_memory+0x1d1/0x520
        [ 5053.836873]  [<ffffffffb195189a>] __alloc_pages_slowpath+0xc2a/0xc60
        [ 5053.836873]  [<ffffffffb1951a6b>] __alloc_pages_nodemask+0x19b/0x1d0
        [ 5053.864438]  [<ffffffffb199fcda>] alloc_pages_vma+0xaa/0x2c0
        [ 5053.864438]  [<ffffffffb19904cc>] __read_swap_cache_async+0x11c/0x1a0
        [ 5053.864438]  [<ffffffffb1990571>] read_swap_cache_async+0x21/0x50
        [ 5053.864438]  [<ffffffffb19906a9>] swapin_readahead+0x109/0x1c0
        [ 5053.864438]  [<ffffffffb1979c1e>] do_swap_page+0x41e/0x7a0
        [ 5053.864438]  [<ffffffffb197b2f0>] handle_mm_fault+0xbe0/0x15f0
        [ 5053.864438]  [<ffffffffb197a756>] ? handle_mm_fault+0x46/0x15f0
        [ 5053.864438]  [<ffffffffb18476b4>] ? __do_page_fault+0x154/0x4b0
        [ 5053.864438]  [<ffffffffb18477bb>] __do_page_fault+0x25b/0x4b0
        [ 5053.864438]  [<ffffffffb1847a1c>] do_page_fault+0xc/0x10
        [ 5053.864438]  [<ffffffffb1de8872>] page_fault+0x22/0x30
        [ 5053.912805] Mem-Info:
        [ 5053.913761] active_anon:138 inactive_anon:198 isolated_anon:35
        [ 5053.913761]  active_file:749 inactive_file:698 isolated_file:0
        [ 5053.913761]  unevictable:0 dirty:25 writeback:177 unstable:0
        [ 5053.913761]  slab_reclaimable:3197 slab_unreclaimable:23932
        [ 5053.913761]  mapped:938 shmem:5 pagetables:766 bounce:0
        [ 5053.913761]  free:10476 free_pcp:118 free_cma:0
        [ 5053.923485] Node 0 active_anon:424kB inactive_anon:856kB active_file:3244kB inactive_file:2456kB unevictable:0kB isolated(anon):140kB isolated(file):0kB mapped:3752kB dirty:100kB writeback:708kB shmem:20kB shmem_thp: 0kB shmem_pmdmapped: 0kB anon_thp: 0kB writeback_tmp:0kB unstable:0kB pages_scanned:558 all_unreclaimable? no
        [ 5053.933761] Node 0 DMA free:7384kB min:384kB low:480kB high:576kB active_anon:0kB inactive_anon:0kB active_file:0kB inactive_file:0kB unevictable:0kB writepending:0kB present:15992kB managed:15908kB mlocked:0kB slab_reclaimable:4kB slab_unreclaimable:12kB kernel_stack:0kB pagetables:0kB bounce:0kB free_pcp:0kB local_pcp:0kB free_cma:0kB
        [ 5053.940135] lowmem_reserve[]: 0 1805 1805 1805
        [ 5053.941529] Node 0 DMA32 free:35140kB min:44668kB low:55832kB high:66996kB active_anon:280kB inactive_anon:476kB active_file:2664kB inactive_file:2580kB unevictable:0kB writepending:256kB present:2080640kB managed:1848448kB mlocked:0kB slab_reclaimable:12784kB slab_unreclaimable:95716kB kernel_stack:3344kB pagetables:3064kB bounce:0kB free_pcp:672kB local_pcp:0kB free_cma:0kB
        [ 5053.947993] lowmem_reserve[]: 0 0 0 0
        [ 5053.950277] Node 0 DMA: 4*4kB (UE) 1*8kB (E) 2*16kB (UE) 1*32kB (E) 2*64kB (UE) 2*128kB (UE) 1*256kB (E) 1*512kB (E) 2*1024kB (UE) 2*2048kB (ME) 0*4096kB = 7384kB
        [ 5053.954936] Node 0 DMA32: 2023*4kB (UMEH) 1046*8kB (UMEH) 456*16kB (UMEH) 222*32kB (UME) 69*64kB (UMEH) 6*128kB (UMH) 0*256kB 0*512kB 0*1024kB 0*2048kB 0*4096kB = 36044kB
        [ 5053.963618] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=1048576kB
        [ 5053.968195] Node 0 hugepages_total=0 hugepages_free=0 hugepages_surp=0 hugepages_size=2048kB
        [ 5053.971264] 1264 total pagecache pages
        [ 5053.972586] 130 pages in swap cache
        [ 5053.973709] Swap cache stats: add 881258, delete 881128, find 273948/484264
        [ 5053.975139] Free swap  = 2072012kB
        [ 5053.976311] Total swap = 2094076kB
        [ 5053.977528] 524158 pages RAM
        [ 5053.978882] 0 pages HighMem/MovableOnly
        [ 5053.983319] 58069 pages reserved
        [ 5053.984494] 0 pages hwpoisoned
        [ 5053.985759] [ pid ]   uid  tgid total_vm      rss nr_ptes nr_pmds swapents oom_score_adj name
        [ 5053.998690] [  192]     0   192    10057      235      24       3      125             0 systemd-journal
        [ 5054.013092] [  224]     0   224    11150      242      22       3      298         -1000 systemd-udevd
        [ 5054.022654] [  391]   100   391    31820        0      32       4      153             0 systemd-timesyn
        [ 5054.028144] [  413]     0   413    11604        0      27       3      145             0 systemd-logind
        [ 5054.029827] [  415]     0   415    17485        0      38       3      193         -1000 sshd
        [ 5054.031715] [  417]     0   417    62529        0      30       3      249             0 rsyslogd
        [ 5054.033887] [  419]   106   419    11279        0      26       4      119          -900 dbus-daemon
        [ 5054.044374] [  424]     0   424     7409        0      19       3       79             0 cron
        [ 5054.048497] [  430]     0   430     3634        0      11       3       38             0 agetty
        [ 5054.050271] [  434]     0   434     3578        0      12       3       37             0 agetty
        [ 5054.064389] [  565]     0   565    18856        0      41       3      325             0 apache2
        [ 5054.066834] [  576]    33   576    91441        0      71       3      870             0 apache2
        [ 5054.068513] [  580]    33   580    91410        0      71       3      852             0 apache2
        [ 5054.071480] [  741]   105   741    14039        0      30       3      186             0 exim4
        [ 5054.077661] [  752]     0   752    23792       38      49       3      277             0 sshd
        [ 5054.082766] [  754]     0   754    14105        0      32       3      210             0 systemd
        [ 5054.088659] [  755]     0   755    20629        0      42       3      380             0 (sd-pam)
        [ 5054.090871] [  761]     0   761     5302      236      15       3      474             0 bash
        [ 5054.092672] [ 3153]    33  3153    91426        0      71       3      826             0 apache2
        [ 5054.094842] [ 3196]    33  3196   107820        0      72       3     1074             0 apache2
        [ 5054.101404] Out of memory: Kill process 3196 (apache2) score 1 or sacrifice child
        [ 5054.103650] Killed process 3196 (apache2) total-vm:431280kB, anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
        [ 5054.187637] oom_reaper: reaped process 3196 (apache2), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB

**Comment by kazan417** [comment]
What SOCK_TEMPESTA flag does? can it affect memory leak?
nginx of course does not use this flag.
Can we recompile nginx or something else with SOCK_TEMPESTA flag for sure that memory leak not in kernel patch?

**Comment by krizhanovsky** [comment]
@kazan417 I don't think so. We introduced `SOCK_TEMPESTA` in https://github.com/tempesta-tech/tempesta/commit/927a7bc2982d699a9d7ea267360297c8aee29908 just to [eliminate TCP accept queue processing](https://github.com/tempesta-tech/tempesta/issues/145). We even don't export the flag to user space, so this is absolutely Tempesta specific.

**Comment by krizhanovsky** [comment]
The issue is quickly reproducable with `ab` on loopback (ran inside VM) with disabled keep-alive connections:

        # grep PG /proc/tempesta/perfstat; ab -r -c 64 -n 8192 http://192.168.100.4/ >/dev/null 2>&1; grep PG /proc/tempesta/perfstat
        AK DBG PG_tempesta: 350933
        AK DBG PG_tempesta: 361267
        # grep PG /proc/tempesta/perfstat; ab -r -c 64 -n 8192 -k http://192.168.100.4/ >/dev/null 2>&1; grep PG /proc/tempesta/perfstat
        AK DBG PG_tempesta: 361267
        AK DBG PG_tempesta: 361267

`AK DBG PG_tempesta` is a debug patch which prints leaked pages.

**Comment by vladtcvs** [comment]
for testing purposes: For reproducing this bug backend should return enough big answer, otherwise it would be too long to notice memory leak.
